[analyzer] Report violations of the "returns_nonnull" attribute

necto opened the pull request:
Make sure code respects the GNU-extension `__attribute__((returns_nonnull))`.

Extend the NullabilityChecker to check that a function returns_nonnull does not return a nullptr.

This commit also reverts an old hack introduced by 49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f
because it is no longer needed

CPP-4741

necto committed:
[analyzer] Report violations of the "returns_nonnull" attribute

Make sure code respects the GNU-extension
__attribute__((returns_nonnull)).

Extend the NullabilityChecker to check that a function returns_nonnull
does not return a nullptr.

CPP-4741

NagyDonat commented on clang/test/Analysis/nullability.c:
```diff
@@ -10,3 +12,42 @@ void block_arity_mismatch() {
+  // False positive: expected-warning@-1 {{Null returned from a function that is expected to return a non-null value}}
+}
+
+__attribute__((returns_nonnull)) int *passthrough(int *p) {
+  return p; // no-warning: we have no evidence that `p` is null, i.e., violating the contract
+}
```
Consider adding a variant of this where a passthrough function is used in a situation where it's known to return null.
```cpp
__attribute__((returns_nonnull)) int *passthrough2(int *p) {
  return p;
}
void call_with_null(void) {
  passthrough2(NULL);
}
```

necto replied:
Added in 3d5e750aec0d 
However, it did not trigger because of an explicit suppression of the reports in inlined functions:
https://github.com/llvm/llvm-project/commit/49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f

I believe the reasoning for that suppression no longer holds, so I lifted the suppression: 598c574e0703
Do you agree? If so, do you think it should be done in a separate change set?

NagyDonat replied:
If the tests within `nullability.mm` still pass (expect for the over-suppression one), then there is no need to preserve that suppression.

I'd say that it's OK to lift the suppression within this commit, but add a remark like "This commit also reverts an old hack introduced by 49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f because it is no longer needed." to the commit message. (Perhaps putting it into a separate commit would be _slightly_ more elegant, but I don't think that it would produce real practical benefits.)

NagyDonat commented on clang/test/Analysis/nullability.c:
```diff
@@ -10,3 +12,42 @@ void block_arity_mismatch() {
+  int *x = produce_nonnull_ptr();
+  if (!x) {
+    clang_analyzer_warnIfReached();
+    // Incorrect: expected-warning@-1 {{REACHABLE}}
+    // According to produce_nonnull_ptr contract, x cannot be null.
```
```suggestion
    // expected-warning@-1 {{REACHABLE}}
    // TODO: This warning is a false positive, according to the contract of
    // produce_nonnull_ptr, x cannot be null.
```
(Just bikeshedding.)

By the way, do you have plans to eliminate this false positive (perhaps in a follow-up commit)?

necto replied:
We've entered it as a Jira ticket: https://sonarsource.atlassian.net/browse/CPP-5614
But we have no immediate plans to implement it.

NagyDonat commented on clang/test/Analysis/nullability.c:
```diff
@@ -1,4 +1,6 @@
-// RUN: %clang_analyze_cc1 -fblocks -analyzer-checker=core,nullability -Wno-deprecated-non-prototype -verify %s
+// RUN: %clang_analyze_cc1 -fblocks -analyzer-checker=core,nullability,debug.ExprInspection -Wno-deprecated-non-prototype -verify %s
```
Consider eliminating the use of  `-Wno-deprecated-non-prototype` by specifiying the parameter lists of functions. (Note that this is C, so you need `return_t funcname(void)` to specify a function that takes no arguments.)

necto replied:
Done in a9c489f683fc

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/NullabilityChecker.cpp:
```diff
@@ -692,6 +692,14 @@ void NullabilityChecker::checkPreStmt(const ReturnStmt *S,
   NullConstraint Nullness = getNullConstraint(*RetSVal, State);
 
   Nullability RequiredNullability = getNullabilityAnnotation(RequiredRetType);
+  if (const auto *FunDecl = C.getLocationContext()->getDecl();
```
Why isn't this logic added as a part of `getNullabilityAnnotation`? (I'm just curious, I can easily imagine that there is a good reason for that.)

necto replied:
`getNullabilityAnnotation` is concerned only with the type. It is used both for return types and function parameters by multiple checkers.
Handling the function annotation there would mean I have to pass the function declaration (which might not be available in the client of the API), which is only relevant in one case (and maybe another couple in TrustNonnullChecker). That means I would need to complicate the function definition and potentially increase the PR footprint by touching all call sites.

NagyDonat replied:
I see, thanks for the clarification.

NagyDonat reviewed (commented):
Thanks for this improvement, I'm really happy to see this! (I added some inline comments, but they are all minor and tangential.)   

Right now all the `nullability.*` checkers are marked as (ObjC) in [the documentation](https://clang.llvm.org/docs/analyzer/checkers.html#nullability), because (as far as I know) previously they were only supporting the Objective-C style `_Nullable` and `_Nonnull` annotations.

However, as your change shows, the same "backend" can be used to handle different kinds of nullability annotations, which would make these checkers available for more users.

Please update the documentation of the affected checkers (in `clang/docs/analyzer/checkers.rst`) to mark that they are no longer limited to Objective-C.

necto committed:
Update clang/test/Analysis/nullability.c

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

necto committed:
Add test for inlined nullability violation

necto committed:
Undo report suppression for null returned from non-null functions

This removes the additional constraint introduced in 49bd58f1ebe28d97e4949e9c757bc5dfd8b2d72f
because its premise (producing FP if a parameter reclaimed early) seems
to no longer hold, and because it interferes with C coverage.

necto committed:
Remove -Wno-deprecated-non-prototype

necto committed:
Update documentation

mention that NullReturnedFromNonnull now also works for C

necto commented:
> Please update the documentation of the affected checkers (in clang/docs/analyzer/checkers.rst) to mark that they are no longer limited to Objective-C.

Updated in 2d2ab3185a00

NagyDonat commented on clang/docs/analyzer/checkers.rst:
```diff
@@ -588,8 +588,8 @@ Warns when a null pointer is passed to a pointer which has a _Nonnull type.
 
-nullability.NullReturnedFromNonnull (ObjC)
-""""""""""""""""""""""""""""""""""""""""""
+nullability.NullReturnedFromNonnull
+"""""""""""""""""""""""""""""""""""
```
```suggestion
nullability.NullReturnedFromNonnull (C, C++, ObjC)
""""""""""""""""""""""""""""""""""""""""""""""""""
```
This is how other checkers are marked. (Do I understand correctly that `__attribute__((returns_nonnull))` is relevant both under C and C++?)

necto replied:
Yes, it is relevant for both. Thanks for the fix! Applied

NagyDonat commented on clang/test/Analysis/nullability.c:
```diff
@@ -51,3 +54,15 @@ int *cannot_return_null() {
   return p; // no-warning: we have no evidence that `p` is null, i.e., violating the contract
 }
+
+__attribute__((noreturn))
+void exit(int);
```
Is this declaration used somehow?

necto replied:
It is not used, indeed; thanks for spotting it!
It was relevant in one of the working versions of the test case, but not in the committed one. Removed in 41956123

NagyDonat reviewed (commented):
Thanks for the updates!

necto committed:
Update clang/docs/analyzer/checkers.rst

Co-authored-by: Donát Nagy <donat.nagy@ericsson.com>

necto committed:
Remove unused function dclaration

NagyDonat reviewed (approved):
Looks good to me now, thanks :smile:. Feel free to merge this / I can help with merging if you don't have commit access yet.

necto committed:
Merge branch 'main' into az/CPP-4741-returns-nonnull
